### DISCUSSION TITLE
await lock

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#8514] [General]
Now, .Net9 has introduced a dedicated type for `lock`, and I think it is meaningful to supplement the `AsyncLock` API and introduce a compatible `await lock` at this point.

**Comment by siegfriedpammer** [comment]
https://github.com/neosmart/AsyncLock

I think `using` is perfectly fine.

**Reply by koszeggy** [reply]
Seems to be a simple wrapper around a `SemaphoreSlim(1, 1)`. Please note that this is [not exactly the same](https://stackoverflow.com/a/71638844/5114784) as a `lock`.

**Comment by timcassell** [comment]
There is no async lock type in the BCL. The new `lock` was only added for the exact `System.Threading.Lock` type, and was not implemented as a general pattern. I think it should be expanded to a general pattern before opening up `await lock` pattern.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/runtime/issues/28194
